src/autora/state/delta.py

```diff
@@ -0,0 +1,321 @@
+    Function to extend supported datatypes.
+
+    """
+    raise NotImplementedError("`extend` not implemented for %s, %s" % (a, b))
+
+
+@extend.register(list)
+def extend_list(a, b):
+    """
+    Examples:
+        >>> extend([], [])
+        []
+
+        >>> extend([1,2], [3])
+        [1, 2, 3]
+    """
+    return a + b
+
+
+@extend.register(pd.DataFrame)
+def extend_pd_dataframe(a, b):
+    """
+    Examples:
+        >>> extend(pd.DataFrame({"a": []}), pd.DataFrame({"a": []}))
+        Empty DataFrame
+        Columns: [a]
+        Index: []
+
+        >>> extend(pd.DataFrame({"a": [1,2,3]}), pd.DataFrame({"a": [4,5,6]}))
+           a
+        0  1
+        1  2
+        2  3
+        3  4
+        4  5
+        5  6
+    """
+    return pd.concat((a, b), ignore_index=True)
+
+
+def append(a: List[T], b: T) -> List[T]:
+    # TODO: add DOCTESTS
+    return a + [b]
+
+
+@extend.register(np.ndarray)
+def extend_np_ndarray(a, b):
+    """
+    Examples:
+        >>> extend(np.array([(1,2,3), (4,5,6)]), np.array([(7,8,9)]))
+        array([[1, 2, 3],
+               [4, 5, 6],
+               [7, 8, 9]])
+    """
+    return np.row_stack([a, b])
+
+
+@extend.register(dict)
+def extend_dict(a, b):
+    """
+    Examples:
+        >>> extend({"a": "cats"}, {"b": "dogs"})
+        {'a': 'cats', 'b': 'dogs'}
+    """
+    return dict(a, **b)
+
+
+def wrap_to_use_state(f):
+    """Decorator to make target `f` into a function on a `State` and `**kwargs`.
+
+    This wrapper makes it easier to pass arguments to a function from a State.
+
+    It was inspired by the pytest "fixtures" mechanism.
+
+    Args:
+        f:
+
+    Returns:
+
+    Examples:
+        >>> from autora.state.delta import State, Delta
+        >>> from dataclasses import dataclass, field
+        >>> import pandas as pd
+        >>> from typing import List, Optional
+
+        The `State` it operates on needs to have the metadata described in the state module:
+        >>> @dataclass(frozen=True)
+        ... class S(State):
+        ...     conditions: List[int] = field(metadata={"delta": "replace"})
+
+        We indicate the inputs required by the parameter names.
+        The output must be a `Delta` object.
+        >>> from autora.state.delta import Delta
+        >>> @wrap_to_use_state
+        ... def experimentalist(conditions):
+        ...     new_conditions = [c + 10 for c in conditions]
+        ...     return Delta(conditions=new_conditions)
+
+        >>> experimentalist(S(conditions=[1,2,3,4]))
+        S(conditions=[11, 12, 13, 14])
+
+        >>> experimentalist(S(conditions=[101,102,103,104]))
+        S(conditions=[111, 112, 113, 114])
+
+        >>> from autora.variable import VariableCollection, Variable
+        >>> from sklearn.base import BaseEstimator
+        >>> from sklearn.linear_model import LinearRegression
+
+        >>> @wrap_to_use_state
+        ... def theorist(experimental_data: pd.DataFrame, variables: VariableCollection, **kwargs):
```

Comment:
just wondering about the introduction of `experimental_data `vs `experiment_data`. they're both the collection of X's and y's, right?

Reply:
thanks, that was an error. I'll update the naming.